Feature/issue#73 알림 기능 연결

Jeonwoojae opened the pull request:
### ✍️ Description
- 기존 구현한 알림 기능을 다른 팀원들이 구현한 코드에 맞게 수정합니다.

<br>

### 🎯 Key Changes
- [X] 음악 추가 기능과 알림 기능 연결
  - [X] 단위 테스트 작성
  - [X] 통합 테스트 작성
- [ ] 친구 추가 기능과 알림 기능 연결
- [ ] 플레이 리스트 공유 기능과 알림 기능 연결 
  - [ ] 플레이 리스트 음악 저장 기능 수정
  - [ ] 플레이 리스트 
<br>

### 💬 To Reviewers
- 현재 계속해서 코드 수정 중입니다. 일단 돌아가는 것을 목표로 해서 좀 복잡한데 개선할 부분 알려주시면 수정하겠습니다!

Jeonwoojae committed:
feat: member의 artist 좋아요 기능 구현

- 좋아요 정보를 저장하기 위해 매핑 테이블을 생성하여 좋아요 시 매핑 테이블에 데이터 저장

Jeonwoojae committed:
feat: member의 좋아요 한 artist 조회 기능 구현

Jeonwoojae committed:
feat: member의 좋아요 한 artist 삭제 기능 구현

Jeonwoojae committed:
test: 아티스트 좋아요 관련 테스트 코드 작성

Jeonwoojae committed:
test: 아티스트 좋아요 관련 통합 테스트 코드 작성

Jeonwoojae committed:
feat: 아티스트 좋아요 관련 기능 컨트롤러 구현

Jeonwoojae committed:
refactor: artist 폴더 변경

Jeonwoojae committed:
refactor: 알림 기능 CompletableFuture를 사용하도록 변경

- CompletableFuture를 사용함으로 테스트 시 결과를 얻을 때 까지 대기할 수 있다.

Jeonwoojae committed:
refactor: 음악 생성 로직 연결

- Spotify에서 다루는 내용 처럼 Song에서 Track으로 이름 변경
- N:N관계로 좋아요한 사용자를 찾기 위해 쿼리 지정
- 테스트 코드 엔티티 생성 정보 수정, 매핑테이블 직접 저장

Jeonwoojae committed:
refactor: 친구 추가, 플레이리스트 공유 이벤트 변경

- 실제 구현한 Member 정보를 넣도록 변경

Jeonwoojae committed:
refactor: 사용하지 않는 임시 객체들 삭제

- Temp 관련 임시 객체들 삭제

Jeonwoojae committed:
refactor: 역할에 따라 Track 관련 코드 폴더 분리

Jeonwoojae committed:
refactor: 아티스트 좋아요 기능 서비스 코드로 분리

- db에 저장하여 아티스트와 사용자 간의 관계 형성

Jeonwoojae committed:
fix: 배치 작업 오류 해결

- 가져온 데이터를 dto로 변환하는 과정에서 자동으로 setter를 찾게 되는데 setter를 사용하지 않기 위해 fieldSet에서 dto 생성하도록 구현

Jeonwoojae committed:
fix: 매핑 테이블 Lazy loading으로 변경

- 연관관계의 엔티티가 필요할 때만 가져올 수 있도록 변경

Jeonwoojae committed:
feat: artist와 track, member와 artist 연관된 객체 조회 기능 구현

- fetch join과 @EntityGraph 방식 사용하며 사용방법 비교

Jeonwoojae committed:
refactor: 레이어 역할에 따라 repository에서 직접 가져오지 않고 service에서 가져오도록 변경

- artist명으로 좋아요 한 사용자 가져오는 기능 service 코드로 분리
- track으로 artist들 가져오는 기능 service 코드로 분리

minju412 commented on src/main/java/com/petitapetit/miml/domain/artist/domain/MemberArtist.java:
```diff
@@ -0,0 +1,56 @@
+        this.artist = artist;
+    }
+
+    public Artist getArtist() {
```
`@Getter`가 선언되어 있기 때문에 생략 가능할 것 같습니다! :)

Jeonwoojae replied:
네 맞습니다!
어노테이션을 사용하였기 때문에 해당 메서드들은 생략했습니다. 감사합니다👍

minju412 commented on src/main/java/com/petitapetit/miml/domain/artist/dto/ArtistDto.java:
```diff
@@ -0,0 +1,27 @@
+	@NoArgsConstructor
+	@Getter
+	@Builder
+	public static class Response {
```
`LikeResponse`로 명명한다면 더욱 직관적일 것 같습니다!

Jeonwoojae replied:
네 그냥 `Response`보다는 의미 있는 `LikeResponse`가 더 직관적일 것 같습니다.
감사합니다!👍

minju412 commented on src/main/java/com/petitapetit/miml/domain/artist/service/ArtistService.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public Set<Member> findMembersByLikedArtistNames(List<Artist> artists) {
+        Set<Member> likedMembers = memberRepository.findByLikedArtistNames(artists.stream().map(Artist::getName).collect(
+                Collectors.toList()));
+        return likedMembers;
```
```suggestion
       return memberRepository.findByLikedArtistNames(artists.stream()
         .map(Artist::getName)
         .collect(Collectors.toList())
       );
```
변수를 생성하지 않고 바로 리턴할 수 있을 것 같아요!

Jeonwoojae replied:
네 변수 이름이 사용할 필요가 없기 때문에 인라인화 해도 좋을 것 같습니다!

minju412 commented on src/main/java/com/petitapetit/miml/domain/member/model/Member.java:
```diff
@@ -34,4 +38,12 @@ public class Member {
 	private OAuth2Provider provider;
 	private String providerId;
+	@OneToMany(mappedBy = "member")
+	@Builder.Default
```
필드가 이미 기본값으로 초기화되어 있는데, `@Builder.Default` 어노테이션을 사용하신 이유가 궁금합니다!

Jeonwoojae replied:
기본값으로 초기화 하고 빌더를 사용했을 때 해당 객체의 필드는 null로 초기화 되지 않아 문제가 발생했었습니다.
해당 문제를 해결하기 위해 빌더에서의 기본값을 지정하기 위해 해당 어노테이션을 사용해 초기화 할 수 있었습니다!

minju412 commented on src/main/java/com/petitapetit/miml/domain/member/repository/MemberRepository.java:
```diff
@@ -11,4 +15,6 @@
 public interface MemberRepository extends JpaRepository<Member, Long> {
 	Optional<Member> findByProviderAndProviderId(OAuth2Provider oAuth2Provider, String providerId);
+	@Query("select m from Member m join m.likedArtists ma where ma.artist.name in :artistNames")
+	Set<Member> findByLikedArtistNames(@Param("artistNames") List<String> artistNames);
```
메서드 매개변수 이름과 JPQL 쿼리 매개변수 이름이 같아 `@Param("artistNames")`를 생략할 수 있을 것 같은데, 명시적으로 나타내기 위해 생략하지 않으신 걸까요?!

Jeonwoojae replied:
네 저는 명시적으로 보이는게 더 좋아서 생략하지 않습니다!

minju412 commented on src/main/java/com/petitapetit/miml/domain/member/service/MemberService.java:
```diff
@@ -0,0 +1,26 @@
+
+    public void likeArtist(Member member, Long artistId) {
+        Artist artist = artistRepository.findById(artistId)
+                .orElseThrow(()->new NoSuchElementException());
```
위에서는 비슷한 상황에 `IllegalArgumentException`을 사용하셨던 것 같은데 여기서는 `NoSuchElementException`을 사용하신 기준이 궁금합니다!

Nine-JH replied:
저도 `NoSuchElementException`은 `Iterator` 타입에 더 어울려보이네요. 😄

Nine-JH replied:
추가적으로 이것도 중복된 코드이므로 묶을 수 있을 것 같아보여요~

Jeonwoojae replied:
여러 Exception를 선언해서 오히려 무슨 에러인지 알기 힘들 것 같다는 의견에 동의합니다..😢
엔티티를 가져오는 메서드들을 모두 분리했고, db에서 데이터를 가져오는 경우에 발생하는 에러는 모두 `NoEntityFoundException`을 반환하도록 변경하겠습니다!

minju412 commented on src/main/java/com/petitapetit/miml/domain/notification/entity/TrackAddedNotification.java:
```diff
@@ -0,0 +1,43 @@
+    @ElementCollection
+    private List<String> songArtist;
+
+    public static TrackAddedNotification from(Track song, Member user){
```
정적 팩토리 메서드의 네이밍 컨벤션에 따라 `of` 혹은 객체를 생성하는 목적을 명확히 하고자 한다면 `createTrackAddedNotification` 등으로 변경하는 것은 어떨까요?
- `from` : 하나의 매개 변수를 받아서 객체를 생성
- `of` : 여러개의 매개 변수를 받아서 객체를 생성

Jeonwoojae replied:
현재 상황에서 `createTrackAddedNotification`으로 메서드를 변경하게 된다면 실제 메서드 사용 시 `TrackAddedNotification.createTrackAddedNotification(song,user)`형태가 되어 오히려 이해하기 힘들 것 같다고 생각합니다. 따라서 지금 여러 개의 매개 변수를 받아 객체를 생성하기 때문에 `from`에서 `of`로 메서드 네이밍을 변경하도록 하겠습니다. 감사합니다👍

minju412 commented on src/main/java/com/petitapetit/miml/domain/notification/service/AsyncNotificationEventHandler.java:
```diff
@@ -0,0 +1,30 @@
+
+    @TransactionalEventListener(classes = TrackAddedEvent.class, phase = TransactionPhase.AFTER_COMMIT)
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public void handleSongEvent(TrackAddedEvent event) {
```
메서드 명을 `handleTrackAddedEvent`로 변경하여 더욱 명확하게 표현할 수 있을 것 같아요 :)

Jeonwoojae replied:
네 `Track`에서 `Song`으로 변경하며 놓친 변수나 메서드 명이 많은 것 같습니다.
해당 부분 외에도 통일된 단어 `Track`를 사용하도록 하고,
어떤 이벤트를 처리할 수 있는지 메서드 명으로 알 수 있게 변경하겠습니다👍

minju412 commented on src/main/java/com/petitapetit/miml/domain/notification/service/AsyncNotificationEventHandler.java:
```diff
@@ -0,0 +1,30 @@
+    public void handleSongEvent(TrackAddedEvent event) {
+        CompletableFuture<Void> result = notificationEventHandler.handleSongEvent(event);
+        try {
+            result.get();
```
23번째 줄에 선언된 `result` 변수를 제거하고 try문 안에서 `notificationEventHandler.handleSongEvent(event).get();`으로 한번에 처리할 수 있을 것 같아요!

+) `join()` 대신 `get()`을 사용하신 이유가 있을까요? [CompletableFuture get()과 join()의 차이점](https://luna-archive.tistory.com/32)

Jeonwoojae replied:
해당 객체로 하는 작업이 없어서 한번에 처리할 수 있겠네요!
get()을 사용했던 이유는 첨부 해 주신 내용처럼 인터럽트가 가능하기 때문에 사용했습니다. 알림과 메일을 생성하면서 예외 사항이 발생하지 않을까 생각했기 때문입니다. 하지만 질문 주신 내용을 다시 고민 해 보니 예외는 발생할 수 있지만 이것을 외부로 던지는게 괜찮을까 고민되기도 하네요.. 좋은 고민거리를 주셔서 감사합니다👍

Nine-JH replied:
좋은 고민거리네요! 그런데 정작 NotificationEventHandler 에서는 인터럽트를 컨트롤하는 로직이 없어보이네요 🧐

minju412 reviewed (approved):
고생하셨습니다!! 👍👍

Jeonwoojae committed:
refactor: 사용하지 않는 메서드 삭제

- getter 사용으로 get 메서드 선언 삭제

Jeonwoojae committed:
refactor: Artist 정보 담는 dto명 변경

- Respnse보다 직관성 있는 LikeResponse로 변경

Jeonwoojae committed:
refactor: 아티스트 명으로 좋아요 한 사용자를 찾는 메서드 인라인화

- 변수를 사용하지 않기 때문에 인라인화 한다.

Jeonwoojae committed:
feat: 친구 추가 요청 알림기능 연동

- 친구 추가 요청 시 알림을 생성하여 다룬다.

Jeonwoojae committed:
feat: 플레이리스트 노래 등록 로직 구현

- 플레이리스트와 노래 사이를 관리하는 TrackPlayList를 사용하여 연결해준다.

Jeonwoojae committed:
feat: 플레이리스트에 있는 노래 삭제 로직 구현

- 플레이리스트와 노래 사이를 관리하는 TrackPlayList를 찾아 삭제하여 연결을 해제한다.

Jeonwoojae committed:
fix: 연결 테이블 요소 찾는 코드 수정

- spring data jpa가 자동으로 찾아주는 필드에 맞게 메서드명 수정

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/artist/service/ArtistService.java:
```diff
@@ -0,0 +1,65 @@
+    @Transactional
+    public void likeArtist(Long userId, Long artistId) {
+        Member user = memberRepository.findById(userId)
+                .orElseThrow(() -> new IllegalArgumentException("사용자를 찾을 수 없습니다: " + userId));
```
중복되는 코드들을 묶을 수 있을 것 같아요 😄

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/artist/domain/MemberArtistRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface MemberArtistRepository extends JpaRepository<MemberArtist, Long> {
+    @EntityGraph(attributePaths = {"member", "artist"})
+    Optional<MemberArtist> findByMember_MemberIdAndArtist_Id(Long memberId, Long artistId);
```
메서드 네이밍 방식이 독특한데 JPA 표준 스펙이 `_` 으로 나누는 건가요?

Jeonwoojae replied:
IDE에서 해당 이름으로 추천 해주길래 필요한 요소를 알기 쉬워서 사용했었는데 해당 부분에 대해서 더 알아봤습니다!
[Spring Data JPA 2.6.10 문서](https://docs.spring.io/spring-data/jpa/docs/2.6.10/reference/html/#repositories.query-methods.query-property-expressions)

공식 문서에 따르면 찾을 속성을 수동적으로 정의하기 위해 `_`로 나눌 수 있어서 내부 요소가 모호하게 정의될 경우 `_`가 사용될 수 있다고 합니다. 하지만 기본적으로는 카멜케이스를 사용하기 때문에 모호하지 않은 지금은 카멜케이스로 변경하는 것이 표준 스펙에는 맞습니다!

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/member/model/Member.java:
```diff
@@ -34,4 +38,12 @@ public class Member {
+	@Builder.Default
+	private Set<MemberArtist> likedArtists = new HashSet<>();
+
+	public void likeArtist(MemberArtist memberArtist) {
+		this.likedArtists.add(memberArtist);
+		memberArtist.setMember(this);
+	}
```
`likeArtist()`와 `unlikeArtist()` 메서드는 Member가 아닌 MemberArtist에서 호출하는 것이 어울려보여요
이유는 다음과 같은데요
`Member` 도메인 관점에서는 `MemberArtist`을 꼭 알아야 하나? 라는 생각이 드네요. 뭔가 직접적인 연관성이 있어 보이지 않아보여요. 😄 `MemberArtist` 입장을 볼까요?
```java
public class MemberArtist {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    private Member member;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "artist_id")
    private Artist artist;
}
```
`Member`가 꼭 알아야 하는 정보죠? 이런 관점에서 `MemberArtist`에서 해당 아티스트를 찜하거나 취소하는게 문맥상 더 어울려 보이는 것 같아요

PS :  추가적으로 MemberArtist에서 정적 팩토리 메서드를 잘 사용하면 굳이 객체를 생성 코드와 `likeArtist()`와 같은 연관관계를 매핑해주는 메서드를 둘 다 작성할 필요 없이 합칠 수 있을 것 같아요 😄

Jeonwoojae replied:
해당 내용에 대해서 고민하고 다시금 알게된 내용으로 이번 발표를 진행했습니다!
좋은 고민점 알려주셔서 감사합니다😆

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/notification/event/FriendRequestedEvent.java:
```diff
@@ -9,7 +9,7 @@ public class FriendRequestedEvent {
     private final String requestedUserName;  // 친구 요청 받은 유저 이름
 
-    public FriendRequestedEvent(TempUser currentUser, TempUser requestedUser) {
+    public FriendRequestedEvent(Member currentUser, Member requestedUser) {
         this.currentUserName = currentUser.getEmail();
         this.requestedUserName = requestedUser.getEmail();
     }
```
단지 username 정보만 필요하다면, Entity를 넘겨주는 대신 username만 넘겨주는건 어떨까요?
그리고 착각일 수 있는데 `getEmail()` 이 아닐 수 있겠다는 생각도 드네요 😅

Jeonwoojae replied:
모든 Entity 정보가 아닌 필요한 정보만 넘겨서 해당 이벤트를 생성하도록 변경하겠습니다.
해당 이벤트 정보로 메일을 보내기 때문에 `getEmail()`은 맞지만 이벤트 내부 변수 이름이 `UserName`으로 되어있어 알기 어려우셨던 것 같습니다.
해당 부분도 헷갈리지 않게 각 변수명을  `~UserEmail` 형식으로 변경하겠습니다.

Nine-JH reviewed (commented):
고생하셨습니다! PR Review가 너무나도 늦어버렸네요. 죄송합니다 😢 
몇가지 코멘트를 남겨드렸어요!

Jeonwoojae committed:
fix: 테스트 에러 수정

- 테스트에서 배치 옵션 제거
- TrackAddedEvent에서 LAZY 옵션으로 artist 이름을 받아오지 못하는 문제를 이벤트 생성 시 artist 이름들을 저장하도록 변경
- Artist와 Track을 연결하는 작업을 trackService에서 하도록 변경하여 책임 분리

Jeonwoojae committed:
refactor: 객체 가져오는 메서드 분리

- 자주 사용되는 객체를 가져오는 메서드를 분리
- 예외를 Argument 예외가 아닌 Entity를 찾을 수 없는 에러로 변경

Jeonwoojae committed:
refactor: jpa 속성 표현식 변경

- jpa 쿼리 메서드명 작성 시 표준 명명 규칙에 따라 카멜케이스로 변경

Jeonwoojae committed:
test: 플레이리스트에서 음악 추가, 삭제 테스트

- 단위 테스트를 진행하여 mocking을 통해 save나 delete메서드를 호출하는지 확인

Jeonwoojae committed:
refactor: 좋아요 연결하는 메서드 변경

- MemberArtist에서 정적 팩토리 메서드를 사용하여 관계 객체를 생성하고 매핑하도록 변경

Jeonwoojae committed:
refactor: 좋아요 취소 시 검색 로직 변경

- 사용자와 노래가 존재하는지 먼저 확인 후에 연관 관계가 존재하는지 확인하고 취소 작업을 수행하도록 변경
- cancelLike() 메서드가 업데이트 된 연관 객체를 반환하여 해당 객체만 삭제될 수 있도록 변경
- MemberService의 likeArtist 과정에서 세션 정보를 잃지 않기 위해 @Transactional 추가

Jeonwoojae committed:
fix: 존재하지 않는 클래스 참조 삭제

- TempUser 참조 삭제

Jeonwoojae committed:
test: 통합 테스트 시 DirtiesContext 사용하지 않도록 변경

- DirtiesContext를 사용하지 않기 위해 테스트 과정엣 사용했던 객체 정보만 테스트 종료 시 db에서 삭제
- 인스턴스 변수명을 내부에서 선언하여 삭제할 수 있도록 수정
- 테스트 코드 초기화를 위한 코드 deleteByArtist, deleteByTrack, deleteByUserEmail 작성

Jeonwoojae committed:
feat: 음악 조회기능 구현

- Pageable을 사용하여 음악 리스트를 페이지 단위로 가져올 수 있다.

Jeonwoojae committed:
refactor: 정적 팩도리 메서드명 변경

- 기존 from은 하나의 매개변수를 받아 생성할 때 사용하는 것으로 지금은 Member와 Track을 받아 생성하기 때문에 of로 변경

Jeonwoojae committed:
refactor: event, notification 간 모호한 표현 변경

- 정적 팩토리 메서드 네이밍을 기본 사용에 맞게 매개변수에 따라 of, from 사용
- UserName을 Email과 혼동하여 사용하지 않도록 변수 명 구분하도록 수정
- 필요한 데이터만 event로 전달하여 notification까지 사용하도록 수정
- 메서드나 변수에서 단어 song을 track으로 변경

Jeonwoojae committed:
refactor: track으로 단어 표현 통일

- 메서드나 변수에서 단어 song을 track으로 변경

minju412 commented:
고생하셨습니다! 👍👍 코드를 살펴봤는데, 깔끔하게 변경해 주셔서 추가적으로 리뷰 드릴 부분은 없는 것 같아요!
특히 커밋 메시지의 바디가 잘 적혀있어서 이해하기 좋았습니다! 저도 더욱 세세히 작성할 수 있도록 해야겠네요 :)
